homeassistant/components/alexa/smart_home.py

```diff
@@ -1276,6 +1354,152 @@ def async_api_previous(hass, config, request, entity):
 
 
+def api_error_temp_range(request, temp, min_temp, max_temp, unit):
+    """Create temperature value out of range API error response.
+
+    Async friendly.
+    """
+    temp_range = {
+        'minimumValue': {
+            'value': min_temp,
+            'scale': API_TEMP_UNITS[unit],
+        },
+        'maximumValue': {
+            'value': max_temp,
+            'scale': API_TEMP_UNITS[unit],
+        },
+    }
+
+    msg = 'The requested temperature {} is out of range'.format(temp)
+    _LOGGER.info(msg)
+    return api_error(
+        request,
+        error_type='TEMPERATURE_VALUE_OUT_OF_RANGE',
+        error_message=msg,
+        payload={'validRange': temp_range},
+    )
+
+
+def temperature_from_object(temp_obj, to_unit, interval=False):
+    """Get temperature from Temperature object in requested unit."""
+    from_unit = TEMP_CELSIUS
+    temp = float(temp_obj['value'])
+
+    if temp_obj['scale'] == 'FAHRENHEIT':
+        from_unit = TEMP_FAHRENHEIT
+    elif temp_obj['scale'] == 'KELVIN':
+        # convert to Celsius if absolute temperature
+        if not interval:
+            temp -= 273.15
+
+    return convert_temperature(temp, from_unit, to_unit, interval)
+
+
+@HANDLERS.register(('Alexa.ThermostatController', 'SetTargetTemperature'))
+@extract_entity
+@asyncio.coroutine
+def async_api_set_target_temp(hass, config, request, entity):
+    """Process a set target temperature request."""
+    unit = entity.attributes[CONF_UNIT_OF_MEASUREMENT]
+    min_temp = entity.attributes.get(climate.ATTR_MIN_TEMP)
+    max_temp = entity.attributes.get(climate.ATTR_MAX_TEMP)
+
+    data = {
+        ATTR_ENTITY_ID: entity.entity_id
+    }
+
+    payload = request[API_PAYLOAD]
+    if 'targetSetpoint' in payload:
+        temp = temperature_from_object(
+            payload['targetSetpoint'], unit)
+        if temp < min_temp or temp > max_temp:
+            return api_error_temp_range(
+                request, temp, min_temp, max_temp, unit)
+        data[ATTR_TEMPERATURE] = temp
+    if 'lowerSetpoint' in payload:
+        temp_low = temperature_from_object(
+            payload['lowerSetpoint'], unit)
+        if temp_low < min_temp or temp_low > max_temp:
+            return api_error_temp_range(
+                request, temp_low, min_temp, max_temp, unit)
+        data[climate.ATTR_TARGET_TEMP_LOW] = temp_low
+    if 'upperSetpoint' in payload:
+        temp_high = temperature_from_object(
+            payload['upperSetpoint'], unit)
+        if temp_high < min_temp or temp_high > max_temp:
+            return api_error_temp_range(
+                request, temp_high, min_temp, max_temp, unit)
+        data[climate.ATTR_TARGET_TEMP_HIGH] = temp_high
+
+    yield from hass.services.async_call(
+        entity.domain, climate.SERVICE_SET_TEMPERATURE, data, blocking=False)
+
+    return api_message(request)
+
+
+@HANDLERS.register(('Alexa.ThermostatController', 'AdjustTargetTemperature'))
+@extract_entity
+@asyncio.coroutine
```

Comment:
We've migrated to `async`/`await` syntax. Please drop this decorator and prefix async methods with `async` like `async def async_api_adjust_target_temp`. Replace any `yield from` with `await`.